.github/workflows/flash_analysis.yml

```diff
@@ -30,10 +36,22 @@ jobs:
         run: git config --system --add safe.directory '*'
 
       - name: Build Target
-        run: make ${{ matrix.target }}
+        # Use special flash-analysis target if present. Otherwise fall back to the default target.
+        run: |
+          if make list_config_targets | grep ${{ matrix.target }}_flash-analysis; then
```

Comment:
The main reason for including this is to ensure that FLASH analysis does not fail on old branches. GH actions behave very strangely here.

As can be seen in this example, after updating the script on main, it failed on an old branch: https://github.com/alexcekay/PX4-Autopilot/actions/runs/12144503657/job/33865529881
The reason it failed is that the target _flash-analysis_ does not yet exist on the old branch.

I thought this would not be a problem since GH actions use the .yml revision on the branch. According to the workflow file used for this run, it seems so: https://github.com/alexcekay/PX4-Autopilot/actions/runs/12144503657/workflow
(as can be seen, the __default_ target is still in the workflow file).

However, the run uses the revision from main, which uses the new _flash-analysis_ target.

This is also the reason why FLASH analysis does not fail on this PR. As can be seen in the build steps of this run (https://github.com/PX4/PX4-Autopilot/actions/runs/12181677791/job/33978880303?pr=24072), it uses the _flash-analysis_ target and the _default_ target. This also confuses bloaty (the reason why the FLASH comment below contains unlogical changes).

If both (before and after PR) are compiled with the same target, it works fine, as can be seen here: https://github.com/alexcekay/PX4-Autopilot/pull/14

Reply:
I can't trace down the first two links. They have no branch/pr attached, but the last one does have a branch attached, but I can't figure out which commit it's supposed to be running.

The GitHub actions runner picks up the yml files from the revision you are checking out, and it can't see anything else, so this shouldn't be a problem.

IMHO, You should leave the fix out, and when/if it fails, we should report it to GitHub support. We need to be able to trust GitHub Actions with this.

Do you see any test runs using the wrong yml in this list? https://github.com/PX4/PX4-Autopilot/actions/workflows/flash_analysis.yml?query=

Reply:
Maybe this example can show the situation better. I created a PR in my repo for testing: https://github.com/alexcekay/PX4-Autopilot/pull/17

As you can see the FLASH Analysis works well for the first two commits. Then it does not work anymore for the third commit, despite the commit just removing an empty line in an unrelated file. The FLASH Analysis does however fail due to not finding the target _px4_fmu-v5x_flash-analysis_t_. This comes from a change I did to main (!) https://github.com/alexcekay/PX4-Autopilot/commit/f840d45cee326af9953c9674f488ae0f049dd732

Reply:
Update: So I invested quite some time now into understanding what is going on here. The answer can be found here:
https://github.com/actions/checkout/issues/426#issuecomment-789985604
When a GH action runs on a PR it does **not** simply checkout the PR and use that. It actually creates a merge commit (merges the base branch into the PR) and uses that.

This is also the reason why suddenly workflow steps from the main branch appear. As another (simple example) I created the following: https://github.com/alexcekay/action-playground/pull/1
As can be seen the "Output" step of the "Overwrite" workflow prints something different for the commits despite nothing being changed in the workflow file.

With this knowledge I can however remove the special handling that is currently present, as the new _flash_analysis_ target will be present in old PRs due to the "merge commit" logic.

The only thing that needs to be changed now (because that was always wrong) is not to diff against _github.event.pull_request.base.sha_ because this one is the base branchs head at the time of the PR creation (see https://github.com/orgs/community/discussions/59677#discussioncomment-10782359). This is however not correct as a diff basis due to the merge commit logic. If the main branch added new stuff it will get used for the "after elf" due to the merge commit logic. The "before elf" will not have those main changes due to the PR being created before the main change. To avoid that I would use the _github.event.pull_request.base.ref_. This should be _main_ most of the times. Then we diff the _main_ against the merge commit (PR change + main) and get a diff of the change :)